File: src/main/scala/crusader728/leetcode/bfs/BinaryTreeRightSideView199.scala
package scala.crusader728.leetcode.bfs

import scala.crusader728.leetcode.TreeNode

object BinaryTreeRightSideView199 {
  def rightSideView(root: TreeNode): List[Int] = {
    if(root == null) {
      Nil
    } else {
      val m = collection.mutable.Map.empty[Int, List[TreeNode]]
      @annotation.tailrec
      def bfs(level: List[TreeNode], depth: Int): Unit = {
        if(level.isEmpty) {
          ()
        } else {
          m += (depth -> level)
          val next = for {
            n <- level
            children <- List(n.right, n.left).filter(node => node != null)
          } yield children
          bfs(next, depth + 1)
        }
      }

      bfs(List(root), 0)
      (0 to m.keySet.max).map(h => m(h).head.value).toList
    }
  }
}

File: src/main/withJava/crusader728/leetcode/bfs/BinaryTreeRightSideView199.java
package withJava.crusader728.leetcode.bfs;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

public class BinaryTreeRightSideView199 {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> result = new ArrayList<>();
        if(root == null) {
            return result;
        } else {
            Deque<TreeNode> queue = new ArrayDeque<>();
            queue.offer(root);
            int count = -1;
            while(!queue.isEmpty()) {
                int size = queue.size();
                count++;
                for(int i = 0; i < size; ++i) {
                    TreeNode poll = queue.poll();
                    if(result.size() < count + 1) {
                        result.add(poll.val);
                    } else {
                        result.set(count, poll.val);
                    }
                    if(poll.left != null) {
                        queue.offer(poll.left);
                    }
                    if(poll.right != null) {
                        queue.offer(poll.right);
                    }
                }
            }
            return result;
        }
    }

    private static final class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
    }
}
